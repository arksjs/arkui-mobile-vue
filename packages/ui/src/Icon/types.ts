export type IconData = any

export interface IconProps {
  icon: IconData
  spin?: boolean
  size?: number | string
  width?: number | string
  height?: number | string
  color?: string
}
