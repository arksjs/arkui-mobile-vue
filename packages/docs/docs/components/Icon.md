# Icon 图标

<CodeDemo name="Icon">

<<< @/../../demo/src/components/Basic/Icon/index.vue

</CodeDemo>

## Import

```js
import { AkIcon } from 'arkui-mobile-vue'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

## Props

| 属性   | 类型                | 默认值 | 必填 | 说明                                           |
| ------ | ------------------- | ------ | ---- | ---------------------------------------------- |
| icon   | string \| Component |        | 是   | 设置图标的名称（Sprite）或者传入自定义图标组件 |
| size   | number              | 24     | 否   | icon 的大小                                    |
| width  | number              |        | 否   | icon 的宽度                                    |
| height | number              |        | 否   | icon 的高度                                    |
| color  | string              |        | 否   | icon 的颜色                                    |

## 自定义 SVG 图标

### Vite 导入 SVG

`Vite` 开发者可以通过 [rollup-plugin-svg-sprites](https://github.com/godxiaoji/rollup-plugin-svg-sprites) 来将 SVG 作为 Vue 组件导入。

`vite.config.js`:

```js
import svgSprites from 'rollup-plugin-svg-sprites'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    vue(),
    svgSprites({
      vueComponent: true,
      exclude: ['node_modules/**']
    })
  ]
})
```

在单文件中：

```vue
<template>
  <MyIcon />
  // or:
  <ak-icon :icon="MyIcon">
</template>

<script setup>
import MyIcon from './my-icon.svg?vueComponent'
</script>
```

TypeScript 支持和更多用法可以查看[文档](https://github.com/godxiaoji/rollup-plugin-svg-sprites)。

### 批量导入 SVG

对于一个批量文件夹形式的图标集，可以通过 [svg-sprites-cli](https://github.com/godxiaoji/svg-sprites-cli) 转为 js。

#### 1. 安装 svg-sprites-cli

```sh
npm i -D svg-sprites-cli
```

#### 2. 通过指令执行：

```sh
svgjs -d ./icons -o path/to/lib/svg.js
```

假设 icons 目录下有：

```
- icons
  - Tab
    - home.svg
    - find.svg
    - user.svg
  - loading.svg
```

默认规则转出来的 id 为 `icon-homeTab`, `icon-findTab`, `icon-userTab`, `icon-loading`。

#### 3. 在入口中引入 js：

`main.ts`:

```js
import { createApp } from 'vue'
import App from './App.vue'
import router from './router'

import 'path/to/lib/svg.js'

createApp(App).use(router).mount('#app')
```

#### 4. Icon 组件通过展示

```vue
<ak-icon icon="icon-homeTab" />
```

### 附录：示例图标库

组件示例使用了一套图标库，涵盖了一些日常使用图标，具体效果和使用方式可以查看 [@arksjs/ui-icons](https://github.com/arksjs/arkui-mobile-vue/tree/main/packages/icons)。
