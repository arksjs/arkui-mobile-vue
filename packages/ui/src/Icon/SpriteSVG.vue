<template>
  <svg>
    <use :xlink:href="'#' + iconName"></use>
  </svg>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  name: 'ak-sprite-icon',
  props: {
    iconName: {
      type: String
    }
  }
})
</script>
