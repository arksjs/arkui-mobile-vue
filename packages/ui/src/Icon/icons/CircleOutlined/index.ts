export { default as default } from './CircleOutlined.vue'
