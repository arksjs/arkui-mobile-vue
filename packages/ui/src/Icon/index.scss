@import '../style/var.scss';
@import '../style/animation.scss';

.ak-icon {
  --ak-icon-size: 24px;
  --ak-icon-color: #{$title-color};

  display: inline-block;
  width: var(--ak-icon-size);
  height: var(--ak-icon-size);
  overflow: hidden;
  line-height: 0;
  fill: var(--ak-icon-color);

  &.spin {
    @include rotate-360-animation(800ms);
  }
}
