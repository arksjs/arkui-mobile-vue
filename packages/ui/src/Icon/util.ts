import { type CSSProperties, getNumber } from '../helpers'

export const getIconStyles = (props: {
  size?: number | string
  width?: number | string
  height?: number | string
  color?: string
}) => {
  const styles: CSSProperties = {}

  props.color && (styles['--ta-icon-color'] = props.color)

  if (getNumber(props.width) > 0 && getNumber(props.height) > 0) {
    styles.width = props.width + 'px'
    styles.height = props.height + 'px'
  } else if (getNumber(props.size) > 0) {
    styles['--ta-icon-size'] = props.size + 'px'
  }

  return styles
}
