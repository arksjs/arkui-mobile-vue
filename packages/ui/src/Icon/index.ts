import { withInstall } from '../helpers/with-install'
import Icon from './Icon.vue'

export { Icon }
export const install = withInstall(Icon)
export default Icon
