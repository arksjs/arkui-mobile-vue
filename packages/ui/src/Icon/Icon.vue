<template>
  <component
    :is="component"
    class="ta-icon"
    :class="{ spin }"
    :style="styles"
    :iconName="iconName"
  >
  </component>
</template>

<script lang="ts">
import { computed, defineComponent, type PropType } from 'vue'
import SpriteSVG from './SpriteSVG.vue'
import { colorValidator, iconValidator, isSvgComponent } from '../helpers'
import type { IconData } from './types'
import { getIconStyles } from './util'

export default defineComponent({
  name: 'ta-icon',
  props: {
    icon: {
      type: [String, Object] as PropType<IconData>,
      validator: iconValidator,
      required: true
    },
    spin: {
      type: Boolean,
      default: false
    },
    size: {
      type: [Number, String],
      default: 0
    },
    width: {
      type: [Number, String],
      default: 0
    },
    height: {
      type: [Number, String],
      default: 0
    },
    color: {
      type: String,
      validator: colorValidator
    }
  },
  setup(props) {
    const styles = computed(() => getIconStyles(props))
    const component = computed(() =>
      isSvgComponent(props.icon) ? props.icon : SpriteSVG
    )
    const iconName = computed(() =>
      isSvgComponent(props.icon) ? null : props.icon
    )

    return {
      iconName,
      styles,
      component
    }
  }
})
</script>
