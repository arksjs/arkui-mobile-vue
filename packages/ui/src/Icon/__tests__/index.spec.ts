import { describe, expect, test } from 'vitest'
import { mount } from '@vue/test-utils'
import { Icon } from '@/Icon'
import CircleOutlined from '@/Icon/icons/CircleOutlined'
import { markRaw } from 'vue'

const icon = 'strIcon'

describe('Icon', () => {
  test('snapshot', () => {
    const wrapper = mount(Icon, {
      props: {
        icon
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('prop icon is string', () => {
    const wrapper = mount(Icon, {
      props: {
        icon
      }
    })

    expect(wrapper.find('use').attributes('href')).toBe(`#${icon}`)
  })

  test('prop icon is Component', () => {
    const wrapper = mount(Icon, {
      props: {
        icon: markRaw(CircleOutlined)
      }
    })

    expect(wrapper.findComponent(CircleOutlined)).toBeTruthy()
  })

  test('prop size', () => {
    const wrapper = mount(Icon, {
      props: {
        icon,
        size: 40
      }
    })

    expect(wrapper.attributes('style')).toContain('--ta-icon-size: 40px;')
  })

  test('prop color', () => {
    const wrapper = mount(Icon, {
      props: {
        icon,
        color: '#6667AB'
      }
    })

    expect(wrapper.attributes('style')).toContain('--ta-icon-color: #6667AB;')
  })
})
