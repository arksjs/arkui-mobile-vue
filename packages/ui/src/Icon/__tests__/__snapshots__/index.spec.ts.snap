// Vitest Snapshot v1

exports[`Icon > snapshot 1`] = `
"<svg class=\\"ak-icon\\">
  <use xlink:href=\\"#strIcon\\"></use>
</svg>"
`;
